strawberryfields/compilers/gaussian_merge.py

```diff
@@ -0,0 +1,357 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains a compiler that merges Gaussian operations into their symplectic forms,
+in a Gaussian and non-Gaussian circuit."""
+
+import strawberryfields.program_utils as pu
+
+from .fock import Fock
+from .gaussian_unitary import GaussianUnitary
+
+
+def get_qumodes_operated_upon(op):
+    """
+    Helper function that returns list of integers, which are the qumode indexes that are operated by op
+    """
+    return [reg.ind for reg in op.reg]
+
+
+def get_op_name(op):
+    """
+    Helper function that obtains the string name of an operation
+    """
+    return op.op.__class__.__name__
+
+
+class GaussianMerge(Fock):
+    """Compiler for that merges Gaussian operations into Gaussian Transforms in a hybrid program
+
+
+    As a result, the Gaussian operations that this compiler returns are :class:`~.ops.GaussianTransform`,
+    and :class:`~.ops.Displacement`. Meanwhile, non-Gaussian operations remain unchanged.
+
+    **Example:**
+
+    Consider the following Strawberry Fields program, compiled using the `'gaussian_merge'` compiler:
+
+    .. code-block:: python3
+
+        from strawberryfields.ops import Xgate, Zgate, BSgate, Rgate
+        import strawberryfields as sf
+
+        circuit = sf.Program(1)
+        with circuit.context as q:
+            Rgate(0.6)|q[0]
+            Rgate(0.2)|q[0]
+            Kgate(0.4)|q[0]
+            Rgate(0.1)|q[0]
+            Rgate(0.6)|q[0]
+            Dgate(0.01)|(q[0])
+
+        compiled_circuit = circuit.compile(compiler="gaussian_merge")
+
+    We can now print the compiled circuit, consisting of
+    :class:`~.GaussianTransform`, :class:`~.Dgate` and the Non-Gaussian :
+    >>> compiled_circuit.print()
+    GaussianTransform([[ 0.6967 -0.7174]
+                       [0.7174  0.6967]]) | (q[0])
+    Kgate(0.4)|q[0]
+    GaussianTransform([[ 0.7648 -0.6442]
+                   [0.6442  0.7648]]) | (q[0])
+    Dgate(0.01, 0) | (q[0])
+    """
+
+    short_name = "gaussian_merge"
+    primitives = {
+        # meta operations
+        "All",
+        "_New_modes",
+        "_Delete",
+        # state preparations
+        "Ket",
+        # measurements
+        "MeasureFock",
+        "MeasureHomodyne",
+        # single mode gates
+        "Dgate",
+        "Sgate",
+        "Rgate",
+        "Vgate",
+        "Kgate",
+        # two mode gates
+        "BSgate",
+        "CKgate",
+        "S2gate",
+    }
```

Comment:
I'm curious if it makes sense to include `GaussianTransformation` as a primitive? That way, it will avoid the `GaussianTransformation` from being needlessly decomposed and 're-merged'.

It might also make sense to include all decompositions here, with the exception of the Graph ones.

Reply:
Agreed not sure why I haven't added it